core/server/src/main/java/alluxio/master/journal/CheckpointManager.java

```diff
@@ -0,0 +1,145 @@
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.journal;
+
+import alluxio.Constants;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.util.UnderFileSystemUtils;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * A class which manages the checkpoint for a journal. The {@link #updateCheckpoint(String)} method
+ * will update the journal's checkpoint file to a specified checkpoint, and
+ * {@link recoverCheckpoint()} will recover from any failures that may occur during
+ * {@link #updateCheckpoint(String)}.
+ *
+ * The checkpoint updating process goes
+ * <pre>
+ * 1. Write a new checkpoint named checkpoint.data.tmp
+ * 2. Rename checkpoint.data to checkpoint.data.backup.tmp
+ * 3. Rename checkpoint.data.backup.tmp to checkpoint.data.backup
+ * 4. Rename checkpoint.data.tmp to checkpoint.data
+ * 5. Delete completed logs
+ * 6. Delete checkpoint.data.backup
+ * </pre>
+ */
+public final class CheckpointManager {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  /** The UFS where the journal is being written to. */
+  private final UnderFileSystem mUfs;
+  /**
+   * Absolute path to the checkpoint file. This is where the latest checkpoint is stored. During
+   * normal operation (when not writing a new checkpoint file), this is the only checkpoint file
+   * that exists. If this file exists and there is no {@link #mTempBackupCheckpointPath}, it plus
+   * all completed logs, plus the active log, should represent the full state of the master.
+   */
+  private final String mCheckpointPath;
+  /**
+   * Absolute path to the backup checkpoint file. The latest checkpoint is saved here while renaming
+   * the temporary checkpoint so that we can recover in case the rename fails. If this file and
+   * {@link #mCheckpointPath} both exist, {@link #mCheckpointPath} is the most up to date checkpoint
+   * and {@link #mBackupCheckpointPath} should be deleted.
+   */
+  private final String mBackupCheckpointPath;
+  /**
+   * Absolute path to the temporary backup checkpoint file. This path is used as an intermediate
+   * rename step when backing up {@link #mCheckpointPath} to {@link #mBackupCheckpointPath}. As long
+   * as this file exists, it supercedes mCheckpointPath as the most up to date checkpoint file.
+   */
+  private final String mTempBackupCheckpointPath;
+  /**
+   * A journal writer through which this checkpoint manager can delete completed logs when the
+   * checkpoint is updated.
+   */
+  private final JournalWriter mWriter;
+
+  /**
+   * Creates a new instance of {@link CheckpointManager}.
+   *
+   * @param ufs the under file system holding the journal
+   * @param the directory for the journal
```

Comment:
NIT: the javadoc params seem out of date with the actual parameters of the constructor.